doc: Introduce a hardware model v2 migration guide (#15963)
nrfconnect/sdk-nrf: doc/nrf/config_and_build/hwmv2_migration.rst

> .. _hwmv1_to_v2_migration:

Reviewer: This document should go in the migration folder

Reviewer 2: I agree.

Migrating to the current hardware model
#######################################

> Starting with |NCS| version 2.7.0, the hardware model, which is to say the
> manner in which SoCs and boards are named, defined and used, has been overhauled. This
> section describes how to migrate from the hardware model used in |NCS| versions
> 2.6.0 and earlier to the current one.
> 
> ::note The current hardware model was introduced shortly after the relase of
> 3.6.0 in the Zephyr Project.

Reviewer: ```suggestion
The *hardware model* is the manner in which SoCs and boards are named, defined, and used.
Starting with |NCS| version 2.7.0, the hardware model has been overhauled.

This section describes how to migrate from the hardware model used in |NCS| versions 2.6.0 and earlier (commonly referred to as "hardware model v1")  to the current one (commonly referred to as "hardware model v2").

.. note::
   The current hardware model was introduced in the Zephyr Project shortly after the release of 3.6.0.
```

> The current hardware model adds support for features that had been identified as important
> over the years. Among them:
> 
> - Support for multi-core, multi-arch AMP (Asymmetrical Multi Processing) SoCs
> - Support for multi-SoC boards
> - Support for reusing the SoC and board Kconfig trees outside of the Zephyr and
>   NCS build system
> - Support for advanced use cases with :ref:`zephyr:sysbuild`
> - Removal of all existing arbitrary and inconsistent uses of Kconfig and folder
>   names

Reviewer: ```suggestion
The current hardware model adds support for features like the following:

* Support for multi-core, multi-architecture AMP (Asymmetrical Multi Processing) SoCs.
* Support for multi-SoC boards.
* Support for reusing the SoC and board Kconfig trees outside of the Zephyr and |NCS| build system.
* Support for advanced use cases with :ref:`zephyr:sysbuild`.
* Removal of all existing arbitrary and inconsistent uses of Kconfig and folder names.
```

> More information about the rationale, development and concepts behind the new
> model can be found in the `Zephyr original issue <51831>`_, the
> `Zephyr original Pull Request <50305>`_ and, for a complete set of changes
> introduced, the `hardware model v2 commit`_.
> 
> Some non-critical features, enhancements and improvements of the new hardware
> model are still in development. Check the
> `Zephyr hardware model v2 enhancements issue <69546>`_ for a complete list.

Reviewer: ```suggestion
For more information about the rationale, development, and concepts behind the new model, see the `Zephyr original issue <https://github.com/zephyrproject-rtos/zephyr/issues/51831>`_, the `Zephyr original Pull Request <https://github.com/zephyrproject-rtos/zephyr/pull/50305>`_,.
For the complete set of changes introduced, consult the `hardware model v2 commit <https://github.com/zephyrproject-rtos/zephyr/commit/commit_hash>`_.

Some non-critical features, enhancements, and improvements of the new hardware model are still in development. Check the `Zephyr hardware model v2 enhancements issue <https://github.com/zephyrproject-rtos/zephyr/issues/69546>`_ for a complete list.
```

> For |NCS| users that also have out-of-tree boards, the transition from the previous hardware model to the current one (commonly
> referred to as "hardware model v2") is optional, because |NCS| provides
> full backwards compatibility for any existing out-of-tree boards.

Reviewer: ```suggestion
For |NCS| users that also have out-of-tree boards, the transition from hardware model v1 to hardware model v2 is optional, because |NCS| provides full backwards compatibility for any existing out-of-tree boards.
```

> If you choose not to convert the out-of-tree board, you will however get a
> warning when building, informing you that the board definition is in the format
> of the previous hardware model.

Reviewer: ```suggestion
If you choose not to convert the out-of-tree board, you will receive a warning when building, informing you that the board definition is in the format compatible with hardware model v1.
```

> If you choose instead to convert your board from the previous to the current hardware model, we
> recommend first reading through the :ref:`zephyr:hw_support_hierarchy` and
> subsequent sections to understand the model in detail. You
> can then use the `ncs-example-application conversion commit`_ as an example on
> how to port a simple board. Additionally, a `conversion script`_ is available
> and works reliably in many cases (though multi-core SoCs may not be handled
> entirely). Finally, the `hardware model v2 commit`_ contains the full conversion
> of all existing boards from the old to the current model, so you can use it as a
> complete conversion reference.

Reviewer: ```suggestion
If you choose to convert your board from hardware model v1 to v2, consult the following pages:

* :ref:`zephyr:hw_support_hierarchy` and the subsequent sections to understand the model in detail. 
* Use the `ncs-example-application conversion commit`_ as an example on how to port a simple board.
* Additionally, a `conversion script`_ is available and works reliably in many cases (though multi-core SoCs may not be handled entirely).
* The `hardware model v2 commit`_ contains the full conversion of all existing boards from the old to the current model, so you can use it as a complete conversion reference.
```

.. _Zephyr original issue: https://github.com/zephyrproject-rtos/zephyr/issues/51831
.. _Zephyr original Pull Request: https://github.com/zephyrproject-rtos/zephyr/pull/50305
.. _Zephyr hardware model v2 enhancements issue: https://github.com/zephyrproject-rtos/zephyr/issues/69546
.. _hardware model v2 commit: https://github.com/nrfconnect/sdk-zephyr/commit/8dc3f856229ce083c956aa301c31a23e65bd8cd8
.. _ncs-example-application conversion commit: https://github.com/nrfconnect/ncs-example-application/commit/f9f2da24041a7b1923e49893cca912482c138aae

> .. _conversion script: https://github.com/nrfconnect/sdk-zephyr/blob/main/scripts/utils/board_v1_to_v2.py

Reviewer: These should all go in links.txt
